pkg/functions/client_test.go

```diff
@@ -1888,3 +1904,132 @@ func TestClient_RunRediness(t *testing.T) {
 		t.Fatalf("err on job stop. %v", err)
 	}
 }
+
+// TestClient_BuildCleanFingerprint ensures that when building a Function the
+// source controlled state is not modified (git would show no unstaged changes).
+// For example, the image name generated when building should not be stored
+// in function metadata that is checked into source control (func.yaml).
+func TestClient_BuildCleanFingerprint(t *testing.T) {
+
+	// Create a temporary directory
+	root, cleanup := Mktemp(t)
+	defer cleanup()
+
+	// create new client
+	client := fn.New()
+
+	f := fn.Function{Root: root, Runtime: TestRuntime, Registry: TestRegistry}
+	ctx := context.Background()
+
+	// init a new Function
+	f, err := client.Init(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// NOTE: Practically one would initialize a git repository, check the source code
+	// and compare that way. For now this only compares fingerprint before and after
+	// building Function
+
+	// get fingerprint before building
+	hashA, _, err := fn.Fingerprint(root)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Build a function
+	if f, err = client.Build(ctx, f); err != nil {
+		t.Fatal(err)
+	}
+
+	// Write to disk because client.Build just stamps (writing is handled in its caller)
+	if err = f.Write(); err != nil {
+		t.Fatal(err)
+	}
+
+	// compare fingerprints before and after
+	hashB, _, err := fn.Fingerprint(root)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if hashA != hashB {
+		t.Fatal("just building a Function resulted in a dirty function state (fingerprint changed)")
+	}
+}
+
+// Test_RemoveInvokedOnOldFunction checks that Remover was invoked after a
+// subsequent redeploy to a new namespace.
+// specifically: deploy to 'nsone', redeploy to 'nstwo' with --namespace flag
+// overriding the ns and expect the remover to be invoked for old Function in ns 'nsone'
+func TestClient_RemoveInvokedOnOldFunction(t *testing.T) {
+	// Create a temporary directory
+	root, cleanup := Mktemp(t)
+	defer cleanup()
+
+	nsOne := "nsone"
+	nsTwo := "nstwo"
+
+	// A mock remover which fails if the name from the func.yaml is not received.
+	remover := mock.NewRemover()
+
+	client := fn.New()
+	// initialize function with namespace defined as nsone
+	_, err := client.Init(fn.Function{Runtime: "go", Root: root,
+		Deploy: fn.DeploySpec{Namespace: nsOne}})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// 	// set deploy command
+	cmd := cmd.NewDeployCmd(cmd.NewTestClient(
```

Comment:
You can just remove the `cmd` here and use the client instance above directly.  This is because we are in the client package, testing only the client.